src/katgpucbf/fgpu/engine.py

```diff
@@ -895,7 +895,7 @@ async def _chunk_send_and_cleanup(
                 chunk.cleanup()
                 chunk.cleanup = None  # Potentially helps break reference cycles
 
-    async def run_transmit(self) -> None:
+    async def run_transmit(self) -> None:  # noqa: C901
```

Comment:
I think this is a sign that we should lift the whole `if dig_total_power is not None: ...` block into a helper function.